core/src/main/java/bio/terra/pearl/core/service/participant/merge/ParticipantMergeService.java

```diff
@@ -0,0 +1,199 @@
+                                   AnswerService answerService,
+                                   KitRequestService kitRequestService,
+                                   ParticipantUserService participantUserService,
+                                   MergeDao mergeDao) {
+        this.portalParticipantUserService = portalParticipantUserService;
+        this.participantTaskService = participantTaskService;
+        this.participantDataChangeService = participantDataChangeService;
+        this.enrolleeService = enrolleeService;
+        this.withdrawnEnrolleeService = withdrawnEnrolleeService;
+        this.surveyResponseService = surveyResponseService;
+        this.answerService = answerService;
+        this.kitRequestService = kitRequestService;
+        this.participantUserService = participantUserService;
+        this.mergeDao = mergeDao;
+    }
+
+    @Transactional
+    public ParticipantUserMerge applyMerge(ParticipantUserMerge merge, DataAuditInfo auditInfo) {
+        for (MergeAction<Enrollee, EnrolleeMerge> enrolleeMerge : merge.getEnrollees()) {
+            applyMergeEnrollee(enrolleeMerge, merge, auditInfo);
+        }
+        portalParticipantUserService.delete(merge.getPpUsers().getSource().getId(), CascadeProperty.EMPTY_SET);
+        participantUserService.delete(merge.getUsers().getSource().getId(), CascadeProperty.EMPTY_SET);
+        return merge;
+    }
+
+    protected void applyMergeEnrollee(MergeAction<Enrollee, EnrolleeMerge> mergeAction, ParticipantUserMerge parentMerge, DataAuditInfo auditInfo) {
+        if (mergeAction.getAction().equals(MergeAction.Action.NO_ACTION)) {
+            // nothing to do
+        } else if (mergeAction.getAction().equals(MergeAction.Action.DELETE_SOURCE)) {
+            // delete the source enrollee
+            deleteMergedEnrollee(mergeAction.getSource(), auditInfo);
+        } else if (mergeAction.getAction().equals(MergeAction.Action.MOVE_SOURCE)) {
+            // move the source enrollee to the target user
+            moveEnrollee(mergeAction.getSource(), parentMerge.getPpUsers().getTarget(), auditInfo);
+        } else if (mergeAction.getAction().equals(MergeAction.Action.MERGE)) {
+            // merge the source enrollee into the target enrollee
+            mergeEnrolleeData(mergeAction.getSource(), mergeAction.getTarget(), mergeAction.getMergePlan(), parentMerge.getPpUsers().getTarget(), auditInfo);
+            deleteMergedEnrollee(mergeAction.getSource(), auditInfo);
+        }
+    }
+
+    protected void deleteMergedEnrollee(Enrollee enrollee, DataAuditInfo auditInfo) {
+        withdrawnEnrolleeService.withdrawEnrollee(enrollee, auditInfo);
+    }
+
+    /** merges data from one enrollee into another.  Does not delete the source enrollee */
+    protected void mergeEnrolleeData(Enrollee source, Enrollee target, EnrolleeMerge mergePlan, PortalParticipantUser targetPpUser, DataAuditInfo auditInfo) {
```

Comment:
where do survey responses get merged together? e.g., if an enrollee has some responses in source and other responses in target? or is that not something we're worried about here? I'm kind of confused because I see things being moved but I can't find where things are actually merged 2 objects to one; what's the difference between `move` and `merge` in this context?

Reply:
Right now, we're not merging responses -- so the merge service treats MERGE the same as MOVE_SOURCE.  I see that's confused you in multiple places -- I'll see if I can come up with a clearer structure.